File: summary/elixir/11_prime_factorization.exs
defmodule PrimeFactors do
  def prime_factors(n) do
    prime_factors(n, 2, [])
  end

  defp prime_factors(1, _, factors), do: Enum.reverse(factors)

  defp prime_factors(n, divisor, factors) when rem(n, divisor) == 0 do
    prime_factors(div(n, divisor), divisor, [divisor | factors])
  end

  defp prime_factors(n, divisor, factors) do
    prime_factors(n, divisor + 1, factors)
  end
end

n = IO.gets("Input a number: ") |> String.trim() |> String.to_integer()
factors = PrimeFactors.prime_factors(n)
IO.puts("Prime factors: #{Enum.join(factors, ", ")}")

File: summary/ts/11_prime_factorization.ts
function primeFactors(n: number): number[] {
  const factors: number[] = [];
  let divisor = 2;

  while (n >= 2) {
    if (n % divisor === 0) {
      factors.push(divisor);
      n = n / divisor;
    } else {
      divisor++;
    }
  }

  return factors;
}

// 使用例
console.log(primeFactors(12)); // [2, 2, 3]
console.log(primeFactors(100)); // [2, 2, 5, 5]
